### PR TITLE
Persistence extensions: allow null serviceId

### PR DESCRIPTION
**Pull request by mherwege**
See https://github.com/openhab/openhab-core/pull/3736#issuecomment-2093140530

### CODE REVIEW

File: bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/extensions/PersistenceExtensions.java [hunk 1]
```diff
@@ -80,7 +80,7 @@ public PersistenceExtensions(@Reference PersistenceServiceRegistry registry,
      * @param item the item to store
      */
     public static void persist(Item item) {
-        internalPersist(item);
+        internalPersist(item, (String) null);
```
Comment (J-N-K) [inline bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/extensions/PersistenceExtensions.java:83]: Why do you need a cast to `String` here?

File: bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/extensions/PersistenceExtensions.java [hunk 2]
```diff
@@ -90,26 +90,22 @@ public static void persist(Item item) {
...
-    public static void persist(Item item, String serviceId) {
+    public static void persist(Item item, @Nullable String serviceId) {
         internalPersist(item, serviceId);
     }
 
-    private static void internalPersist(Item item) {
-        String serviceId = getDefaultServiceId();
-        if (serviceId == null) {
+    private static void internalPersist(Item item, @Nullable String serviceId) {
+        String sId = serviceId == null ? getDefaultServiceId() : serviceId;
```
Comment (J-N-K) [inline bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/extensions/PersistenceExtensions.java:98]:
```suggestion
        String effectiveServiceId = Objects.requireNonNullElse(serviceId, getDefaultServiceId());
```

Also below

**Review by J-N-K — CHANGES_REQUESTED** [review]
Otherwise a nice addition.

**Comment by jimtng** [comment]
Perhaps label this as regression?

**Comment by jimtng** [comment]
Also to ensure that this is honored in the future, could we please add some tests to check?

**Comment by J-N-K** [comment]
I don't think it's a regression. It didn't work before, right?

**Comment by jimtng** [comment]
> I don't think it's a regression. It didn't work before, right?

It worked like this before #3736:
When given null as the serviceId, it would use the default serviceId, the same as when the serviceId is omitted completely.

#3736 removed this and instead, it actually looks up `null` in the PersistenceServiceRegistry and naturally not find it, and as a result, fails the call by returning null. So yes this is very much a regression, although as noted, this behaviour wasn't documented in the javadoc.

File: bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/extensions/PersistenceExtensions.java (hunk 1 above)
Comment (mherwege) [reply bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/extensions/PersistenceExtensions.java]: Was needed when I only half changed the file (version with signature `internalPersist(Item, TimeSeries)`), but no longer needed, so removed.

**Comment by mherwege** [comment]
> When given null as the serviceId, it would use the default serviceId, the same as when the serviceId is omitted completely.

True, it would, but relying on it was dangerous. For some actions it did not work, and for others it gave wrong result as it was never meant to be used like that. For some, you got lucky and it actually did end up giving the right result.

**Review by J-N-K — APPROVED** [review]
Thanks!

**Pull request merged**